### PR TITLE
fix: mock_modules clobbering collections and args false positives

### PR DESCRIPTION
**Pull request by djdanielsson**
## Summary

- Write `mock_modules` / `mock_roles` stubs under `{cache_dir}/ansible-lint-mocks/` instead of `{cache_dir}/collections/`, so offline lint no longer overwrites real modules in `~/.ansible/collections`.
- Register mock paths as **fallback** search paths (append, not prepend) so installed collections take precedence over empty stubs.
- Skip `args` validation when the resolved module is a lint-generated stub (marker-based, same pattern as `.ps1` modules in #3200).
- Never overwrite an existing non-mock module file when writing stubs; warn if legacy stubs remain under the real collections tree.
- Remove lint-generated module stub files during mockings cleanup.

## Problem

With `offline: true` and a large `mock_modules` list, ansible-lint wrote empty stubs (`argument_spec=dict()`) into `cache_dir/collections/...` — the same tree galaxy uses under `~/.ansible`. Those stubs were prepended first on the collections search path, so `ArgsRule` loaded them and reported every real parameter as unsupported (`args[module]` / `(basic.py)`). This also **clobbered** installed collection module files.

## Test plan

- [x] `pytest test/test_mockings.py test/rules/test_args.py test/test_app.py` (32 passed)
- [x] `test_mock_modules_do_not_touch_real_collections_tree` — stubs stay under `ansible-lint-mocks/`
- [x] `test_mock_modules_do_not_overwrite_installed_module` — real files are not overwritten
- [x] `test_args_rule_skips_lint_mock_modules` — playbook with mocked FQCN + `enable_list: [args]` → no `args[module]` violations
- [x] `test_perform_mockings_cleanup_removes_module_stubs` — stub files removed on cleanup

## Related issues

- #5013 — mock stubs under `.ansible/collections` in AH gating workflows
- #5145 — cache collections path ordering (this fix keeps cache-first for real installs only)
- #3200 — precedent for ArgsRule early-return on unvalidatable module paths

## Recovery for affected users

If modules under `~/.ansible/collections/...` start with `# This is a mocked Ansible module generated by ansible-lint`, reinstall affected collections:

```bash
ansible-galaxy collection install <namespace>.<collection> --force
```

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **New Features**
  * Improved offline linting with safer mock modules and roles stored in a dedicated cache.
  * Real installed Ansible modules and collections now take precedence over generated mocks.
  * Added safer handling for mock paths, invalid names, stale stubs, and cleanup.
  * Added support for `container` as a container engine, with documented automatic fallback order.

* **Bug Fixes**
  * Prevented generated mocks from overwriting installed modules or collections.
  * The `args` rule now skips generated empty-specification mocks.

* **Documentation**
  * Expanded configuration, caching, and rule documentation, including legacy cache recovery guidance.

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -27,18 +27,29 @@ the `GITHUB_ACTIONS=true` and `GITHUB_WORKFLOW=...` variables.
...
+You should add project-local `.ansible` / `.cache` folders to `.gitignore`.
+
+!!! warning "Recovering clobbered collections"
+
+    Older ansible-lint releases wrote `mock_modules` stubs directly into
+    `{cache_dir}/collections/`, which could replace real module files when
+    `offline: true` used `~/.ansible`. If modules under
+    `~/.ansible/collections/...` start with
+    `# This is a mocked Ansible module generated by ansible-lint`, reinstall
+    the affected collections with `ansible-galaxy collection install ... --force`.
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply docs/usage.md]: Agreed — fixed in 8b2a9b7. Rewrote the recovery admonition to use a fenced `bash` block for the `ansible-galaxy ... --force` example so MD046 is satisfied.

File: src/ansiblelint/_mockings.py [hunk 2]
```diff
@@ -57,41 +107,49 @@ def _make_module_stub(module_name: str, options: Options) -> None:
...
+                mock_collections_path
                 / "ansible_collections"
                 / namespace
                 / collection
                 / "roles"
                 / role_dir
             )
         else:
-            path = options.cache_dir / "roles" / role_name
+            path = mock_root / "roles" / role_name
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/ansiblelint/_mockings.py]: Valid finding — fixed in 8b2a9b7.

Simple `mock_roles` names are now validated (`^[\w][\w.-]*$`, no `..` / path separators), and all role/module mock paths go through `_safe_path_under_root()` so resolved paths cannot escape `ansible-lint-mocks`. Added `test_mock_roles_reject_path_escape` covering `../../outside`.

**Comment by djdanielsson** [comment]
### CI / CodeRabbit follow-up

Addressed both CodeRabbit review comments in 8b2a9b7:

1. **docs/usage.md (MD046)** — recovery admonition now uses a fenced `bash` block.
2. **mock_roles path escape** — validate role names and constrain resolved paths under `ansible-lint-mocks` (with regression test).

On the previous green tox run, all required tox jobs passed (`lint-pkg-schemas-eco`, py matrix, docs, hook, check). Remaining red checks appear unrelated to this change:

- **codecov/project** — coverage gate / timing (often races while uploads settle)
- **rpm-build:fedora-rawhide-x86_64** — Packit SRPM build, not exercised by these source changes

File: src/ansiblelint/_mockings.py [hunk 3]
```diff
@@ -6,44 +6,127 @@
...
+    if re.match(r"^\w+\.\w+\.\w+$", role_name):
+        namespace, collection, role_dir = role_name.split(".")
+        return _safe_path_under_root(
+            mock_collections_path,
+            "ansible_collections",
+            namespace,
+            collection,
+            "roles",
+            role_dir,
+        )
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/ansiblelint/_mockings.py:113]: Agreed — fixed in 352bfa8.

FQCN module and role mock paths are now built from `mock_root` with `collections` as a path component via `_safe_path_under_root()`, so a symlinked `ansible-lint-mocks/collections` cannot redirect writes outside the mocks tree. Added `test_mock_paths_reject_symlinked_collections_escape`.

File: test/test_mockings.py [hunk 4]
```diff
@@ -95,3 +104,217 @@ def test_short_mock_modules_stub_is_lintable(tmp_path: Path) -> None:
...
+    """Path helper rejects traversal and separator injection."""
+    from ansiblelint._mockings import _safe_path_under_root
+
+    root = tmp_path / "ansible-lint-mocks"
+    root.mkdir()
+
+    assert _safe_path_under_root(root, "roles", "valid_role") is not None
+    assert _safe_path_under_root(root, "roles", "..", "outside") is None
+    assert _safe_path_under_root(root, "roles", "nested/role") is None
+    assert _safe_path_under_root(root, "roles", "") is None
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply test/test_mockings.py:334]: Agreed — fixed in 352bfa8. Extended `test_safe_path_under_root_rejects_invalid_parts` to cover `"."` and Windows-style `\\` path components.

**Comment by shvenkat-rh** [comment]
@coderabbitai review

**Comment by djdanielsson** [comment]
@coderabbit full review

File: docs/usage.md [hunk 5]
```diff
@@ -27,18 +27,32 @@ the `GITHUB_ACTIONS=true` and `GITHUB_WORKFLOW=...` variables.
...
+when not offline, or `~/.ansible` / `ANSIBLE_HOME` when `offline: true`).
+
+Mocked modules and roles from `mock_modules` / `mock_roles` are written under
+`{cache_dir}/ansible-lint-mocks/` so they never overwrite installed collection
+content. That mock tree is added as a **fallback** on the collections/module
+search path: real installed plugins take precedence, and stubs are used only
+when the real plugin is missing (for example offline syntax-check).
+
+Ansible-lint does not automatically clean the cache. If required you can delete
+the cache directory manually; a new cache is created on the next invocation.
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply docs/usage.md]: Valid — with `offline: true`, `{cache_dir}` can be `~/.ansible` / `ANSIBLE_HOME`, so recommending deletion of the whole cache directory was unsafe.

Updated the Caching section in `docs/usage.md` to:
- point mock cleanup at `{cache_dir}/ansible-lint-mocks/`
- allow deleting project-local `{project}/.ansible`
- warn against deleting shared Ansible home when it holds installed roles/collections

Fixed in fe849b4e.

**Comment by shvenkat-rh** [comment]
This pull request resolves issues with mock_modules clobbering installed collections and causing false positives in the args rule. The key changes include:
Key Changes
1. Improved mock module placement: Stubs are now written under {cache_dir}/ansible-lint-mocks/ instead of overwriting real collections in ~/.ansible/collections
2. Fallback search paths: Mock paths are appended (not prepended) to the collections search path, so installed collections take precedence
3. Args rule enhancement: The args rule now skips lint-generated stubs during validation to avoid false positives
4. Safety improvements: Added protections against overwriting real modules and cleaning up legacy stubs
Benefits
- Fixes clobbering of installed collection modules when using offline: true
- Eliminates false args[module] violations for mocked modules
- Improves performance by properly prioritizing installed collections first
- Adds better safeguards for preventing unsafe path operations
- Provides clear recovery instructions for users affected by older versions
Testing
The PR includes comprehensive tests covering:
- Mock module and role creation
- Path validation and safety checks
- Cleanup functionality
- Args rule behavior with mocked modules
- Legacy stub cleanup
- Edge cases around path traversal
This is a well-tested, important fix that improves the reliability of ansible-lint's offline capability while maintaining backward compatibility. The changes are focused, safe, and address real pain points in the current implementation.

**Review by shvenkat-rh — APPROVED** [review]
